src/main/ipfs/ipfs.ts

```diff
@@ -0,0 +1,133 @@
+import { Request } from "./request_json"
+import { Response } from "./response_json";
+import * as net from 'net'
+import { DaemonRequest } from "./classes";
+
+class IPFS {
+    HOST: string = 'localhost';
+    PORT: number = 7777;
+    
+    client: net.Socket;
+
+    constructor() {
+        this.client = new net.Socket();
+        this.init();
+    }
+
+    private init() {
+        this.client.connect(this.PORT, this.HOST, () => {
+            console.log(`Connected to daemon (${this.HOST}:${this.PORT})`);
+        })
+
+        this.client.on('end', () => {
+            console.log("Disconnected from daemon");
+        })
+    }
+
+    /*
+     * Manually exit the connection with the daemon.
+     */
+    exit() {
+        this.client.end();
+    }
+
+    /*
+     * ping the daemon. Mainly used for testing or verifying that the daemon is
+     * running. Returns whether the ping was correctly received.
+     */
+    ping() {
+        return new Promise((resolve, reject) => {
+            var request   = new Request;
+            request.Class = DaemonRequest.PING;
+    
+            if (! this.client.write(request.serialize())) {
+                console.log("Failed to write request to TCP connection");
+                reject("Failed to write request to TCP connection");
+            }
+
+            this.client.on('data', (data) => {
+                var response = this.parseResponse(data);
+
+                console.log(`Class: ${response.Class}`);
+                console.log(`Msg:   ${response.Msg}`);
+
+                resolve(true);
+            })
+
+            this.client.on('error', (err) => {
+                console.log(`Error: ${err}`);
+                reject(err);
+            })
+        });
+    }
+
+    /*
+     * add a file or directory to IPFS. The string returned on success is the
+     * CID of the added content.
+     */
+    add(path: string) {
+        return new Promise((resolve, reject) => {
+            var request   = new Request;
+            request.Class = DaemonRequest.ADD;
+            request.Path  = path;
+
+            if (! this.client.write(request.serialize())) {
```

Comment:
Seems like we do this a bunch -> can we make a small class-level helper function for this?

Reply:
While the code flagged above is that of an old commit, the point applies for the newer commit as well. We seem to be using `    if (!client.write(request.serialize())) {
      reject(new Error('Failed to send request to daemon'));
    }

    client.on('data', (data) => {
      resolve(parseResponse(data).Msg);
    });` in more than one place in a promise. Can we functionalize this?

Reply:
👍

Reply:
Added helper for `client.write`, but leaving `client.on('data', ...)` since this is not constant for all daemon interactions. It depends on the type of interaction.